### PR TITLE
rec: Account for the NSEC(3) denial TTL in expanded wildcard answers

### LINKED ISSUE #9967: signed wildcard expansions might get cached for too long
In DNSSEC, a wildcarded name (say, `*.example.com`) is served as two things together:
* the wildcard, expanded (`hello.example.com`), with the original signature over `*.example.com`
* a NSEC(3) proof showing that `hello.example.com` does not exist and thus the wildcard expansion is legit

All of these records have TTLs, and so do their RRSIGs. The proof, and thus the expansion, is only valid for the lowest of all these TTLs (the TTLs themselves also being capped by signature expiry).

It appears that our Recursor, however, caches wildcards (plus NSEC(3) proofs) for the TTL of the expanded wildcard. If somebody validates downstream from us, we might serve them an expanded wildcard with expired proofs.

Thus, we should limit the caching of expanded wildcards to the lowest TTL involved in the expansion and the proof.

Note from @rgacogne: 

> we should likely take authorityRecs into account here: https://github.com/PowerDNS/pdns/blob/master/pdns/syncres.cc#L3104

### PR DESCRIPTION
**Pull request by rgacogne**
### Short description

We need to take into account the TTL of NSEC(3) proofs in expanded wildcard answers, as well as the expiration time of their signatures.

Fixes #9967.

### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [ ] included documentation (including possible behaviour changes)
- [x] documented the code
- [ ] added or modified regression test(s)
- [x] added or modified unit test(s)

### CODE REVIEW

File: pdns/syncres.cc [hunk 1]
```diff
@@ -2327,9 +2328,32 @@ uint32_t SyncRes::computeLowestTTD(const std::vector<DNSRecord>& records, const
 
     for(const auto& sig : signatures) {
       if (isRRSIGNotExpired(d_now.tv_sec, sig)) {
-        // we don't decerement d_sigexpire by 'now' because we actually want a TTD, not a TTL */
+        // we don't decrement d_sigexpire by 'now' because we actually want a TTD, not a TTL */
         lowestTTD = min(lowestTTD, static_cast<uint32_t>(sig->d_sigexpire));
       }
+      else {
+        /* if the signature IS expired, we should not cache this */
+        lowestTTD = d_now.tv_sec;
```
Comment (rgacogne) [inline pdns/syncres.cc:2336]: This bit might be controversial, especially for backports. I think I should remove it, we already cap the TTL of Bogus records anyway.

File: pdns/recursordist/test-syncres_cc5.cc [hunk 2]
```diff
@@ -814,6 +814,10 @@ BOOST_AUTO_TEST_CASE(test_dnssec_validation_nsec_wildcard)
   BOOST_CHECK_EQUAL(res, RCode::NoError);
   BOOST_CHECK_EQUAL(sr->getValidationState(), vState::Secure);
   BOOST_REQUIRE_EQUAL(ret.size(), 4U);
+  for (const auto& rec : ret) {
+    /* check that we applied the lowest TTL, here this is from the NSEC proving that the exact name did not exist */
+    BOOST_CHECK_LE(rec.d_ttl, 60);
```
Comment (omoerbeek) [inline pdns/recursordist/test-syncres_cc5.cc:819]:
```suggestion
    BOOST_CHECK_LE(rec.d_ttl, 60U);
```

File: pdns/recursordist/test-syncres_cc5.cc [hunk 3]
```diff
@@ -932,6 +936,10 @@ BOOST_AUTO_TEST_CASE(test_dnssec_validation_nsec_wildcard_proof_before_rrsig)
...
```
Comment (omoerbeek) [inline pdns/recursordist/test-syncres_cc5.cc:941]:
```suggestion
    BOOST_CHECK_LE(rec.d_ttl, 60U);
```

File: pdns/recursordist/test-syncres_cc5.cc [hunk 4]
```diff
@@ -1408,6 +1416,10 @@ BOOST_AUTO_TEST_CASE(test_dnssec_validation_nsec3_wildcard)
   BOOST_CHECK_EQUAL(res, RCode::NoError);
   BOOST_CHECK_EQUAL(sr->getValidationState(), vState::Secure);
   BOOST_REQUIRE_EQUAL(ret.size(), 6U);
+  for (const auto& rec : ret) {
+    /* check that we applied the lowest TTL, here this is from the NSEC3 proving that the exact name did not exist (next closer) */
+    BOOST_CHECK_LE(rec.d_ttl, 60);
```
Comment (omoerbeek) [inline pdns/recursordist/test-syncres_cc5.cc:1421]:
```suggestion
    BOOST_CHECK_LE(rec.d_ttl, 60U);
```

**Review by omoerbeek — APPROVED** [review]
Three nits

**Comment by rgacogne** [comment]
> Three nits

Sorry, Otto, I always mess that up :-/

**Pull request merged**